chapters/chap03-sec03.tex

```diff
@@ -36,25 +36,25 @@ \section{\zeth~statement after primitive instantiation}\label{instantiation:stat
+In this section we detail the method by which we encode (resp.~decode) a set of binary strings to (resp.~from) sets of field elements. In the rest of this section, the notion of \emph{packing policy} refers to the set of \emph{packing} and \emph{unpacking} functions.
 
 The set of primary inputs is composed of the input nullifiers, the output commitments, the public values (see~\cite[Section 3.4.3]{zethpaper}) along with the signature hash and the authentication tags for security (malleability fix, see~\cref{appendix:trnm}). The complete description of the public inputs is represented in~\cref{instantiation:eq:primary-inputs}.
 %
 \begin{equation}\label{instantiation:eq:primary-inputs}
     (\indexedset{\priminputs.\nf{i}}{i \in [\jsin]}, \indexedset{\priminputs.\cms{j}}{j \in [\jsout]}, \vin, \vout, \hsig, \indexedset{\priminputs.\htags{i}}{i \in [\jsin]})
 \end{equation}
 
-Consider the subset of primary inputs that are binary strings, that is the nullifiers $\nfs{}$, the public values $\vin$ and $\vout$, the signature hash $\hsig$ and the authentication tags $\htags{}$.
-For each element $x$ in this subset, $\alpha_x = \ceil {\len{x} / \bnFieldBitCap}$, represent the number of field elements to encode all of the element's bits. Let $\beta_x = \floor{\len{x} / \bnFieldBitCap}$ represent the number of field elements which are fully ``filled'' and let $\gamma_x = \len{x} \pmod{\bnFieldBitCap}$ denote the number of remaining bits. For simplicity, as we assume $\zvalueLen < \bnFieldBitCap$, we only define the notation $\gamma_\notev = \zvalueLen$ for the public values $\vin$ and $\vout$.
+The primary inputs that consist of binary strings are: the nullifiers $\nfs{}$, the public values $\vin$ and $\vout$, the signature hash $\hsig$ and the authentication tags $\htags{}$.
+
+For a binary string $x$, let $\alpha_x = \ceil {\len{x} / \bnFieldBitCap}$ be the number of field elements required to completely encode $x$ and let $\beta_x = \floor{\len{x} / \bnFieldBitCap}$ be the number of field elements whose capacity is fully used. Let $\gamma_x = \len{x} \pmod{\bnFieldBitCap}$ be the number of remaining or ``residual'' bits remaining after fully using $\beta_x$ field elements.
 
 \begin{example}\label{instantiation:eg:packing}
-    Let $\FF'$ be a prime field which elements are encoded on $5$ bits.
-%
-    If we take a binary string $A \in \bin^7$, such that: $A = (1111011)$, we get: $\alpha_A = 2,\ \beta_A = 1,\ \gamma_A = 2$.
-    We illustrate the result of packing $A$ to field elements~\cref{instantiation:fig:packingA}
+    Let $\FF'$ be a prime field whose elements have $5$ bits of capacity (that is, field elements can represent numbers between $0$ and at least $2^{5}-1$). Let $A \in \bin^7$ be a binary string of length $7$. Then $\alpha_A = 2$, $\beta_A = 1$, and $\gamma_A = 2$. That is, $A$ can be represented as 2 field elements, or as 1 field element with 2 ``residual'' bits.
```

Comment:
`between $0$ and at least $2^{5}-1$` -> at most (FF' is a prime field, so we may well want to do arithmetic over e.g. 29 - which binary representation is 11101 - and thus obtain a field with 29 elements)

Not sure this sentence, i.e. `(that is, field elements can represent numbers between $0$ and at least $2^{5}-1$)` is necessary here as we just focus on binary string length for the sake of the example.

Reply:
So, F_29 would have "capacity" 4 (since it can represent all values $0$ up to $2^4$, but not all values up to $2^5$), and "bit length" 5.  In any case, I've removed this since we define "capacity" elsewhere.

Reply:
Ah you changed the initial example (in which I didn't talk - on purpose - about "field capacity" as per the implementation).
When I wrote
> "Let $\FF'$ be a prime field which elements are encoded on $5$ bits."

I meant: `\FF' = \FFx{p}` where `p` is prime such that `ceil(log(p,2)) = 5`
Didn't want to add potential confusion with field capacity (i.e. `floor(log(p,2))`) here is was not necessary to illustrate the "packing" mechanism

Reply:
> I meant: \FF' = \FFx{p} where p is prime such that ceil(log(p,2)) = 5

AFAICT, there is no such $p$ which would also be consistent with $\gamma_A = 2$ and the diagram.
TBH, I was confused that the original does not mention the $floor(log(p,2)) = 5$ requirement, despite giving `\alpha`, `\beta`, `\gamma` etc for A.

We could perhaps give a concrete $p$ with 5 bits capacity if you prefer?  And avoid mentioning capacity?

Reply:
> AFAICT, there is no such $p$ which would also be consistent with $\gamma_A = 2$ and the diagram.

I probably overlooked the example, but basically I'm saying:
Take, say \FFx{31} (31 = (11111)_2), now you have a string A = 1111011 which cannot fit in 5 bits, so the 5 first (or last depending on the packing function definition) bits become a field element and the 7 mod 5 = 2 last bits become the 2nd field elements.

> there is no such $p$ which would also be consistent with $\gamma_A = 2$ and the diagram.

I think I see what you mean: in the example above (see first half of comment) I compute gamma as `7 mod 5` instead of `7 mod 4` (the capacity). In which case, yes I agree there is an inconistency. I didn't want to rely on the "capacity" to explain how packing worked in the example but you are right, this is necessary.

> We could perhaps give a concrete $p$ with 5 bits capacity if you prefer? And avoid mentioning capacity?

Sure, we can rephrase it to raise any ambiguity :)

Reply:
> Sure, we can rephrase it to raise any ambiguity :)

I think $p = 41$ is the smallest suitable $p$, so we could replace with something like this?:
```
Consider binary strings $A \in \bin^7$ of length 7, to be encoded over the field $\FFx{41} $.  Field elements ar encoded over 6 bits, and can reliably encode up to $5$ bits.  This gives $\alpha_A = 2$, $\beta_A = 1$, and $\gamma_A = 2$. That is, $A$ can be represented as 2 field elements, or as 1 field element with 2 ``residual'' bits.
```
The above still mentions "Field elements can reliably encode up to $5$ bits" (required in order to give tha alpha, beta, gamma), so may not be any simpler than just saying "capacity".  
TBH, I'm OK either way.

Reply:
Perfect, let's use:
> Consider binary strings $A \in \bin^7$ of length 7, to be encoded over the field $\FFx{41} $. This field has a capacity of $5$ bits which gives $\alpha_A = 2$, $\beta_A = 1$, and $\gamma_A = 2$. That is, $A$ can be represented as 2 field elements, or as 1 field element with 2 ``residual'' bits.

Reply:
updated this example.

Reply:
Great, thanks @dtebbs !